Improve Sentry state pre-initialization

Gudahtt opened the pull request:
## Explanation

Previously the masked state snapshot sent to Sentry would be blank for errors that occured during initialization. Instead we'll now include some basic information in all cases, and a masked copy of the persisted state if it happens after the first time the persisted state is read.

Relates to #20449

## Manual Testing Steps

* Introduce a migration error (e.g. throw an error in a migration), and ensure it gets sent to Sentry with the persisted state object

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

Gudahtt committed:
Improve Sentry state pre-initialization

Previously the masked state snapshot sent to Sentry would be blank for
errors that occured during initialization. Instead we'll now include
some basic information in all cases, and a masked copy of the persisted
state if it happens after the first time the persisted state is read.

danjm commented on app/scripts/background.js:
```diff
@@ -2,9 +2,22 @@
+import LocalStore from './lib/local-store';
+import ReadOnlyNetworkStore from './lib/network-store';
+
+const inTest = process.env.IN_TEST;
```
Do these declarations have to happen before `import { setupInitialStateHooks } from './lib/setup-initial-state-hooks';`? Why is that?

Gudahtt replied:
`platform` and `localStore` are both passed into `setupInitialStateHooks`. The intent here was for that call to happen as early as possible, so that an error later in this file would include the state snapshot in the error report

Gudahtt replied:
Oh I see, the comment is misleading at the moment. I'll update it.

danjm replied:
Okay, that makes sense.

There is another thing I still don't understand though.  Is there a functional difference between the code as it is:
```
import ExtensionPlatform from './platforms/extension';
import LocalStore from './lib/local-store';
import ReadOnlyNetworkStore from './lib/network-store';

const inTest = process.env.IN_TEST;
const localStore = inTest ? new ReadOnlyNetworkStore() : new LocalStore();
const platform = new ExtensionPlatform();

import { setupInitialStateHooks } from './lib/setup-initial-state-hooks';

setupInitialStateHooks({ localStore, platform });
```
And this alternative order of lines:
```
import ExtensionPlatform from './platforms/extension';
import LocalStore from './lib/local-store';
import ReadOnlyNetworkStore from './lib/network-store';
import { setupInitialStateHooks } from './lib/setup-initial-state-hooks';

const inTest = process.env.IN_TEST;
const localStore = inTest ? new ReadOnlyNetworkStore() : new LocalStore();
const platform = new ExtensionPlatform();

setupInitialStateHooks({ localStore, platform });
```

?

Gudahtt replied:
Nope. I can re-order them to put imports first. The goal here was just to get this function call to happen before the imports, so that any errors on import would also be caught.

Though.... I just remembered that imports are hoisted. So this wouldn't even have that effect. Hmm...

Gudahtt replied:
I revised this significantly here: https://github.com/MetaMask/metamask-extension/pull/20491/commits/db52b4e1d2fc590018e1905812bf7c6743923f52

Gudahtt committed:
Add test

Gudahtt committed:
Fix crash when persisted state not yet fetched

Gudahtt committed:
Add descriptions for initial state hooks

Gudahtt committed:
Update comments to reflect recent changes

Gudahtt committed:
Re-order imports to follow conventions

danjm commented:
This PR is looking good so far and the code makes sense.

One other question: how did you discover this problem?

Gudahtt committed:
Move initial state hooks back to module-level

The initial state hooks are now setup at the top-level of their module.
This ensures that they're setup prior to later imports. Calling a
function to setup these hooks in the entrypoint module wouldn't
accomplish this even if it was run "before" the imports because ES6
imports always get hoisted to the top of the file.

The `localStore` instance wasn't available statically, so a new state
hook was introduced for retrieving the most recent retrieved persisted
state.

Gudahtt committed:
Fix error e2e tests
